rfd/0035-aws-node-join.md

```diff
@@ -0,0 +1,263 @@
+---
+authors: Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 35 - Simplified Node Joining for AWS
+
+## What
+
+Teleport nodes running on EC2 instances will be able to join a teleport cluster
+without the need to share any secret token with the node.
+
+Instead, the node will provide either a signed
+[EC2 Instance Identity Document](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html)
+or a signed
+[sts:GetCallerIdentity](https://docs.aws.amazon.com/STS/latest/APIReference/API_GetCallerIdentity.html)
+request that will be used to confirm the AWS account that the EC2 instance is
+running in. With a configured AWS join token on the auth server, the node
+will be allowed to join the cluster based on its AWS account.
+
+## Why
+
+This will make provisioning nodes on EC2 simpler and arguably more secure.
+There is no need to configure static tokens, or to manage the provisioning and
+deployment of dynamic tokens.
+
+## Details
+
+We are currently exploring two ways to implement this, both inspired by [similar
+designs](https://www.vaultproject.io/docs/auth/aws) used by Vault.
+
+The first is the "EC2 Method", which uses signed EC2 Instance Identity Documents
+which EC2 instances can fetch from the AWS
+[IMDSv2](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html).
+
+The second is the "IAM Method", in which the EC2 instance will create and sign
+an HTTPS API request to Amazon's public `sts:GetCallerIdentity` endpoint and
+pass it to the auth server.
+
+### EC2 Method
+
+In place of a join token, nodes will present an EC2 IID (Instance Identity
+Document) to the auth server, with a corresponding PKCS7 signature.
+
+This will be sent along with the other existing parameters in a
+[RegisterUsingTokenRequest](
+https://github.com/gravitational/teleport/blob/d4247cb150d720be97521347b74bf9c526ae869f/lib/auth/auth.go#L1538-L1563).
+
+The auth server will then:
+1. Use the AWS DSA public certificate to check that the PKCS7 signature for the
+   IID is valid (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/verify-pkcs7.html).
+2. Check that the `pendingTime` of the IID, which is the time the instance was
+   launched, is within a 5 minute TTL.
+   - if the node fails to join the cluster during this window, the user can
+     stop and restart the EC2 instance to reset the `pendingTime` (and
+     effectively create a new IID with a new signature)
+3. Check that the AWS join token matches the AWS account in the IID, and the
+   requested Teleport service role.
+4. Check that this IID has not already been used to join the cluster. \*
```

Comment:
whatever heartbeat on AWS is used, identifies the node by instance ID, and includes the signature from the IID. To prevent replay attack, the auth server will first verify the IID signature, getNode(instanceID).

Reply:
If it's running and it has joined the cluster, the heartbeat should be present, so if the node name == instance id

Reply:
For the lookup to be efficient, the node name == instance-id

Reply:
Updated so that the node name will be set to `<aws_account_id>-<aws_instance_id>` (instance IDs are not necessarily unique across accounts) so auth can efficiently check if the node has already joined.

Don't need to store the IID signature as we will block any second join from the same instance.